psi4/src/export_psio.cc

```diff
@@ -56,6 +62,12 @@ void export_psio(py::module &m) {
         .def("tocscan", &PSIO::tocscan,
              "Seek string in binary file. This export is only good for catching None, as returned success object not "
              "exported.")
+	.def("read", &PSIO::read, "Reads data from within a TOC entry from a PSI file", "unit"_a,
+	     "key"_a, "buffer"_a, "size"_a, "start"_a, "end"_a)
+	.def("write", &PSIO::write, "Writes data to a TOC entry in a PSI file", "unit"_a,
+	     "key"_a, "buffer"_a, "size"_a, "start"_a, "end"_a)
+	.def("read_entry", &PSIO::read_entry, "docstring", "unit"_a, "key"_a, "buffer"_a, "size"_a)
+	.def("write_entry", &PSIO::write_entry, "docstring", "unit"_a, "key"_a, "buffer"_a, "size"_a)
```

Comment:
Can you add a test for this? It would be very useful just to demonstrate how this would work to users as well.

Reply:
Absolutely. I am working on it to show its use in the context of combining Psi4's CASSCF with OpenRDM's MCPDFT module. Once I get it done, I will add a test to its corresponding directory in Psi4.

Reply:
I might also design a much simpler test specifically for PSIO class as well. I need to think more about it.

Thanks for the suggestion.

Reply:
Dear Daniel,

I have prepared a very simple test case for the exposed read/write libpsio functions. I could not attach the input file here for some reason. Please find the input file at

https://github.com/SinaMostafanejad/v2rdm_casscf/blob/OpenRDM-interface/libpsio_test.dat

The result however shows that the exposed functions do not work the way they are expected to, possibly because some of the function arguments (such as **char * buffer**) on the C++ side need to be pointers to characters or strings while on the python side, they become immutable objects (strings) and cannot be changed when we try to write in them as buffers.

I am not sure if I am correct in my interpretation of the test result or doing something wrong, but could you please take a look into it and let me know what do you think about the problem?

Thanks. 

Do you have any simple way around this?

Reply:
I don't have anything specific off the top of my head for this, but this kind of issue is one of the main reasons I encouraged a test. See [here](https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html) for more information, see if they have anything on this.

Reply:
I agree with you that any new functionality should be accompanied by a test case to show its use. However, this is libpsio's function that is being exposed to python and therefore, there has been no similar test about it in Psi4 last time i checked. So, I think based on your comment, adding a test would be really useful and I will update you guys on that if I find a solution to this problem.

Reply:
@loriab Do you understand why this test does not work? (https://github.com/SinaMostafanejad/v2rdm_casscf/blob/OpenRDM-interface/libpsio_test.dat)

Reply:
I have hijacked @edeprince3's idea and tried to get around the problem of immutability of the strings in python for these read/write() functions in libpsio by wrapping libpsio's functions inside another which does not use **char *** for buffer arguments

`void PSIO::py_read(size_t unit, const char *key, double* buffer, size_t size, psio_address start, psio_address end) {
    read(unit, key, (char *) buffer, size, start, (psio_address *) &end);
}
`
or 

`void PSIO::py_read(size_t unit, const char *key, double buffer, size_t size, psio_address start, psio_address end) {
    read(unit, key, (char *) &buffer, size, start, (psio_address *) &end);
}
`

It does not solve the problem after exposing these new functions to python. So, the issue might be something else.

Reply:
I can confirm that using **ctypes** module and **ctypes.create_string_buffer()**  and other functions from this module in the above test, results in the same response from libpsio read()/write() functions by not updating the buffer values.

Codes like

```
opdm_c = c_char_p(b'120')
psioh.write_entry(PSIF_V2RDM_D1A, "D1a", repr(byref(opdm_c)), sizeof(opdm_c))

opdm2_c = create_string_buffer(sizeof(opdm_c))
psioh.read_entry(PSIF_V2RDM_D1A, "D1a", repr(opdm2_c), sizeof(opdm2_c))
```
or similar variations like

```
opdm_c = c_char_p(b'120')
psioh.write_entry(PSIF_V2RDM_D1A, "D1a", string_at(addressof(opdm_c)), sizeof(opdm_c))

opdm2_c = create_string_buffer(sizeof(opdm_c))
psioh.read_entry(PSIF_V2RDM_D1A, "D1a", string_at(addressof(opdm2_c)), sizeof(opdm2_c))
```

would not work.